core/version/gear.ex

```diff
@@ -137,22 +142,37 @@ defmodule AntikytheraCore.Version.Gear do
     Enum.each(pairs_not_installed, fn {gear_name, deps} ->
       L.error("#{gear_name} is not installed due to unmatched dependencies: #{inspect(deps)}")
     end)
+
+    num_installable = length(pairs_not_installed) + num_installed
+    # version_upgrade_test starts Antikythera without gears.
+    # Then, we can't check number of installed gears.
+    needs_check = Antikythera.Env.runtime_env() != :local
+
+    is_enough_gears_installed =
+      num_installable != 0 && num_installed / num_installable > @installed_gear_ratio_threshold
+
+    if !needs_check || is_enough_gears_installed do
+      :ok
+    else
+      :error
+    end
   end
 
-  defunpt install_gears_whose_deps_met(
-            pairs :: v[[gear_and_deps_pair]],
-            installed_gears_set :: MapSet.t(),
-            f :: (GearName.t() -> :ok)
-          ) :: [gear_and_deps_pair] do
+  # public for mock
+  defun install_gears_whose_deps_met(
+          pairs :: v[[gear_and_deps_pair]],
+          installed_gears_set :: MapSet.t(),
+          f :: (GearName.t() -> :ok)
+        ) :: {[gear_and_deps_pair], non_neg_integer} do
     if Enum.empty?(pairs) do
-      []
+      {[], MapSet.size(installed_gears_set)}
     else
       {pairs_installable, pairs_not_installable} =
         Enum.split_with(pairs, fn {_, deps} -> MapSet.subset?(deps, installed_gears_set) end)
 
       if Enum.empty?(pairs_installable) do
         # we cannot make progress any more
-        pairs_not_installable
+        {pairs_not_installable, MapSet.size(installed_gears_set)}
       else
         gears_installable = Keyword.keys(pairs_installable)
         Enum.each(gears_installable, f)
```

Comment:
There are two possible reasons why the gear can't be installed. One is the circular reference between gears, the other is the problem of the gear application itself.
This code counts the former problem, but not the latter. We need to check the return value of `f`.

Reply:
I think it should not consider the gears which are failed to install as installed gear, so I fixed this as well.
Since the gears which are failed to install are already reported when `install_gears_whose_deps_met/4` ends, `pairs_not_installable` isn't good variable to save the gears. `install_gears_whose_deps_met/4` returns the number of the gears. Then, `do_install_gears_at_startup/1` uses the number to decide whether it returns :ok.
https://github.com/access-company/antikythera/pull/152/commits/3298e5983cebb6670eaa34b725de484a90d819bb
https://github.com/access-company/antikythera/pull/152/commits/4c5a4d2181c8e42e029b56edd4ca5eccdb785d43